### PR TITLE
Lifter Solenoid final

### PR DESCRIPTION
**Pull request by YoniKiriaty**

### CODE REVIEW

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/ILifter.java [hunk 1]
```diff
@@ -4,18 +4,26 @@
...
 	void goToPosition(Rotation2d position);
-	
+
+	void openSolenoid();
+
+	void closeSolenoid();
+
+	void holdSolenoid();
+
+	void setPowerSolenoid(double powerSolenoid);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/ILifter.java:26]:
```suggestion
	void setPowerToSolenoid(double powerSolenoid);
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java [hunk 2]
```diff
@@ -8,88 +8,90 @@
...
-	
+
 	@Override
 	public void periodic() {
 		super.periodic();
 		lifter.updateInputs(lifterInputs);
-		
-		Logger.processInputs("Lifter/Lifter", lifterInputs);
+
+		Logger.processInputs("Lifter/Lifter",lifterInputs);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java:38]:
```suggestion
		Logger.processInputs("Lifter/Lifter", lifterInputs);
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java [hunk 3]
```diff
@@ -8,88 +8,90 @@
...
 	}
-	
+
 	public boolean isAtPosition(Rotation2d targetPosition) {
 		return Math.abs(targetPosition.getRadians() - lifterInputs.position) < LifterConstants.TOLERANCE.getRadians();
 	}
-	
-	public Rotation2d getPosition() {
+
+	public Rotation2d getPosition(){
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java:78]:
```suggestion
	public Rotation2d getPosition() {
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java [hunk 4]
```diff
@@ -8,88 +8,90 @@
...
-	
+	public void holdSolenoid(){
+		lifter.holdSolenoid();
+	}
+	public void openSolenoid(){
+		lifter.openSolenoid();
+	}
+	public void closeSolenoid(){
+		lifter.closeSolenoid();
+	}
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java:89]:
```suggestion
	public void holdSolenoid() {
		lifter.holdSolenoid();
	}
	public void openSolenoid() {
		lifter.openSolenoid();
	}
	public void closeSolenoid() {
		lifter.closeSolenoid();
	}
```

File: src/main/java/edu/greenblitz/robotName/commands/lifter/ExtendLifter.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Override
+    public void initialize() {
+        lifter.closeSolenoid();
+        timer.start();
+    }
+
+    @Override
+    public void execute() {
+        if (timer.hasElapsed(LifterConstants.SECONDS_TO_CLOSE_SOLENOID)){
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/commands/lifter/ExtendLifter.java:22]:
```suggestion
        if (timer.hasElapsed(LifterConstants.SECONDS_TO_CLOSE_SOLENOID)) {
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java [hunk 6]
```diff
@@ -8,88 +8,90 @@
...
+
 	public Pose3d getLifterPose3d() {
 		return new Pose3d(
 				LifterConstants.ROBOT_RELATIVE_LIFTER_POSITION,
-				new Rotation3d(
-						-lifterInputs.position,
-						0,
-						0
-				).plus(LifterConstants.ROBOT_RELATIVE_LIFTER_ROTATION)
+				new Rotation3d(-lifterInputs.position,0,0).plus(LifterConstants.ROBOT_RELATIVE_LIFTER_ROTATION)
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java:94]:
```suggestion
				new Rotation3d(
				            -lifterInputs.position,
				            0,
				            0
				).plus(LifterConstants.ROBOT_RELATIVE_LIFTER_ROTATION)
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/LifterInputs.java [hunk 7]
```diff
@@ -4,16 +4,22 @@
...
 	public boolean isForwardSwitchPressed;
-	
+
 	public boolean isBackwardSwitchPressed;
+
+	public boolean isOpenSolenoid;
+
+	public double currentSolenoid;
+
+	public double voltageSolenoid;
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/LifterInputs.java:24]:
```suggestion
	public double solenoidCurrent;

	public double solenoidVoltage;
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java [hunk 8]
```diff
@@ -9,58 +11,73 @@
...
 		motor.config(NeoLifterConstants.CONFIG);
-		
-		motor.getReverseLimitSwitch(NeoLifterConstants.BACKWARD_LIMIT_SWITCH_TYPE)
-				.enableLimitSwitch(NeoLifterConstants.IS_BACKWARD_LIMIT_SWITCH_ENABLED);
-		motor.getForwardLimitSwitch(NeoLifterConstants.FORWARD_LIMIT_SWITCH_TYPE)
-				.enableLimitSwitch(NeoLifterConstants.IS_FORWARD_LIMIT_SWITCH_ENABLED);
-		
+
+		motor.getReverseLimitSwitch(NeoLifterConstants.BACKWARD_LIMIT_SWITCH_TYPE).enableLimitSwitch(NeoLifterConstants.IS_BACKWARD_LIMIT_SWITCH_ENABLED);
+		motor.getForwardLimitSwitch(NeoLifterConstants.FORWARD_LIMIT_SWITCH_TYPE).enableLimitSwitch(NeoLifterConstants.IS_FORWARD_LIMIT_SWITCH_ENABLED);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java:24]:
```suggestion
		motor.getReverseLimitSwitch(NeoLifterConstants.BACKWARD_LIMIT_SWITCH_TYPE)
		          .enableLimitSwitch(NeoLifterConstants.IS_BACKWARD_LIMIT_SWITCH_ENABLED);
		motor.getForwardLimitSwitch(NeoLifterConstants.FORWARD_LIMIT_SWITCH_TYPE)
		          .enableLimitSwitch(NeoLifterConstants.IS_FORWARD_LIMIT_SWITCH_ENABLED);
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java [hunk 9]
```diff
@@ -9,58 +11,73 @@
...
+
 	@Override
 	public void goToPosition(Rotation2d position) {
-		motor.getPIDController().setReference(
-				position.getRadians(),
-				CANSparkMax.ControlType.kPosition,
-				NeoLifterConstants.PID_SLOT,
-				NeoLifterConstants.FEED_FORWARD.calculate(position.getRadians())
-		);
+		motor.getPIDController().setReference(position.getRadians(), CANSparkMax.ControlType.kPosition, NeoLifterConstants.PID_SLOT , NeoLifterConstants.FEED_FORWARD.calculate(position.getRadians()));
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java:58]:
```suggestion
		motor.getPIDController().setReference(
		           position.getRadians(), 
		           CANSparkMax.ControlType.kPosition, 
		           NeoLifterConstants.PID_SLOT , 
                   NeoLifterConstants.FEED_FORWARD.calculate(position.getRadians())
        );
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java [hunk 10]
```diff
@@ -9,17 +9,19 @@
...
+
 	private SingleJointedArmSim lifterSimulation;
- 
 	private double appliedOutput;
- 
 	private ProfiledPIDController pidController;
-	
+	DCMotorSim simulationSolenoidMotor;
+	double appliedOutputsSolenoid;
+	double appliedSolenoidOutputs;
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java:23]:
```suggestion
	DCMotorSim simulationSolenoidMotor;
	
	double appliedOutputsSolenoid;
	
	double appliedSolenoidOutputs;
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java [hunk 11]
```diff
@@ -35,49 +37,79 @@ public SimulationLifter() {
...
-	
+
 	@Override
 	public void setVoltage(double voltage) {
-		appliedOutput = MathUtil.clamp(
-				voltage,
-				RobotConstants.SimulationConstants.MIN_MOTOR_VOLTAGE,
-				RobotConstants.SimulationConstants.MAX_MOTOR_VOLTAGE
-		);
+		appliedOutput = MathUtil.clamp(voltage, RobotConstants.SimulationConstants.MIN_MOTOR_VOLTAGE, RobotConstants.SimulationConstants.MAX_MOTOR_VOLTAGE);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java:55]:
```suggestion
		appliedOutput = MathUtil.clamp(
		       voltage, 
	           RobotConstants.SimulationConstants.MIN_MOTOR_VOLTAGE, 
               RobotConstants.SimulationConstants.MAX_MOTOR_VOLTAGE
        );
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java [hunk 12]
```diff
@@ -35,49 +37,79 @@ public SimulationLifter() {
...
+	@Override
+	public void holdSolenoid() {
+		setPowerSolenoid(LifterConstants.POWER_TO_HOLD_SOLENOID);
+	}
+
+	@Override
+	public void setPowerSolenoid(double power) {
+		setVoltageSolenoid(power * RobotConstants.SimulationConstants.BATTERY_VOLTAGE);
+	}
+	public void setVoltageSolenoid(double voltage){
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java:98]:
```suggestion
	public void setVoltageSolenoid(double voltage) {
```

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java (hunk 8 above)
Comment (greenblitz4590) [reply src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java:24]: image

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java (hunk 11 above)
Comment (greenblitz4590) [reply src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java:55]: image

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java (hunk 12 above)
Comment (greenblitz4590) [reply src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java:98]: image

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java [hunk 13]
```diff
@@ -9,65 +11,88 @@
...
+        motor.setVoltage(voltage);
+    }
+
+    @Override
+    public void resetEncoder(Rotation2d position) {
+        motor.getEncoder().setPosition(position.getRadians());
+    }
+
+    @Override
+    public void stopMotor() {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java:47]:
```suggestion
    public void stop() {
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java [hunk 14]
```diff
@@ -9,65 +11,88 @@
...
+        motor.setIdleMode(idleMode);
+    }
+
+    @Override
+    public void goToPosition(Rotation2d position) {
+        motor.getPIDController().setReference(
+                position.getRadians(),
+                CANSparkMax.ControlType.kPosition,
+                NeoLifterConstants.PID_SLOT,
+                NeoLifterConstants.FEED_FORWARD.calculate(position.getRadians()));
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java:62]:
```suggestion
                NeoLifterConstants.FEED_FORWARD.calculate(position.getRadians())
       );
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java [hunk 15]
```diff
@@ -9,75 +9,92 @@
...
+
+    private SingleJointedArmSim lifterSimulation;
+    private double appliedOutput;
+    private ProfiledPIDController pidController;
+    DCMotorSim simulationSolenoidMotor;
+    double appliedOutputsSolenoid;
+    double appliedSolenoidOutputs;
+
+    public SimulationLifter() {
+        lifterSimulation = new SingleJointedArmSim(DCMotor.getNEO(SimulationLifterConstants.NUMBER_OF_MOTORS), SimulationLifterConstants.GEAR_RATIO, SingleJointedArmSim.estimateMOI(LifterConstants.LENGTH_OF_LIFTER, LifterConstants.LIFTER_MASS_KG), LifterConstants.LENGTH_OF_LIFTER, LifterConstants.BACKWARD_LIMIT.getRadians(), LifterConstants.FORWARD_LIMIT.getRadians(), false, LifterConstants.STARTING_ANGLE.getRadians());
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java:26]: spaces..

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java [hunk 16]
```diff
@@ -9,75 +9,92 @@
...
+        pidController = SimulationLifterConstants.SIMULATION_PID;
+
+        simulationSolenoidMotor = new DCMotorSim(DCMotor.getCIM(SimulationLifterConstants.NUMBER_OF_SOLENOID_MOTORS), SimulationLifterConstants.MOTOR_GEARING, SimulationLifterConstants.MOTOR_JKG_METERS_SQUARED);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java:29]: spaces..

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java [hunk 17]
```diff
@@ -9,75 +9,92 @@
...
+    }
+
+    @Override
+    public void setPower(double power) {
+        setVoltage(power * RobotConstants.SimulationConstants.BATTERY_VOLTAGE);
+    }
+
+    @Override
+    public void setVoltage(double voltage) {
+        appliedOutput = MathUtil.clamp(voltage, RobotConstants.SimulationConstants.MIN_MOTOR_VOLTAGE, RobotConstants.SimulationConstants.MAX_MOTOR_VOLTAGE);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java:39]: spaces..

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java [hunk 18]
```diff
@@ -8,88 +8,96 @@
...
-				).plus(LifterConstants.ROBOT_RELATIVE_LIFTER_ROTATION)
-		);
-	}
+    private LifterInputsAutoLogged lifterInputs;
+
+    private Lifter() {
+        lifter = LifterFactory.create();
+        lifterInputs = new LifterInputsAutoLogged();
+        lifter.updateInputs(lifterInputs);
+        setIdleMode(CANSparkMax.IdleMode.kBrake);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java:19]: save in a constant the idlemode

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java [hunk 19]
```diff
@@ -8,88 +8,96 @@
...
+
+    public void setPower(double power) {
+        lifter.setPower(power);
+    }
+
+    public void setVoltage(double voltage) {
+        lifter.setVoltage(voltage);
+    }
+
+    public void resetEncoder(Rotation2d position) {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/Lifter.java:53]:
```suggestion
    public void resetEncoderToPosition(Rotation2d position) {
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java [hunk 20]
```diff
@@ -9,65 +11,89 @@
...
+
+    private GBSparkMax motor;
+    private TalonSRX solenoid;
+
+    public NeoLifter() {
+        motor = new GBSparkMax(NeoLifterConstants.MOTOR_ID, CANSparkLowLevel.MotorType.kBrushless);
+        solenoid = new TalonSRX(NeoLifterConstants.SOLENOID_ID);
+        motor.config(NeoLifterConstants.CONFIG);
+
+        motor.getReverseLimitSwitch(NeoLifterConstants.BACKWARD_LIMIT_SWITCH_TYPE).enableLimitSwitch(NeoLifterConstants.IS_BACKWARD_LIMIT_SWITCH_ENABLED);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java:23]: what happend here..?

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java [hunk 21]
```diff
@@ -9,65 +11,89 @@
...
+        motor.getForwardLimitSwitch(NeoLifterConstants.FORWARD_LIMIT_SWITCH_TYPE).enableLimitSwitch(NeoLifterConstants.IS_FORWARD_LIMIT_SWITCH_ENABLED);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java:24]:
```suggestion
        motor.getForwardLimitSwitch(NeoLifterConstants.FORWARD_LIMIT_SWITCH_TYPE)
                  .enableLimitSwitch(NeoLifterConstants.IS_FORWARD_LIMIT_SWITCH_ENABLED);
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java [hunk 22]
```diff
@@ -9,65 +11,89 @@
...
+
+        motor.setSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, LifterConstants.BACKWARD_LIMIT.getRadians());
+        motor.setSoftLimit(CANSparkMax.SoftLimitDirection.kForward, LifterConstants.FORWARD_LIMIT.getRadians());
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java:27]:
```suggestion
        motor.setSoftLimit(
               CANSparkMax.SoftLimitDirection.kReverse, 
               LifterConstants.BACKWARD_LIMIT.getRadians()
        );
        motor.setSoftLimit(
               CANSparkMax.SoftLimitDirection.kForward, 
               LifterConstants.FORWARD_LIMIT.getRadians()
        );
```

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java [hunk 23]
```diff
@@ -9,75 +9,110 @@
...
+                DCMotor.getNEO(SimulationLifterConstants.NUMBER_OF_MOTORS),
+                SimulationLifterConstants.GEAR_RATIO,
+                SingleJointedArmSim.estimateMOI(
+                        LifterConstants.LENGTH_OF_LIFTER,
+                        LifterConstants.LIFTER_MASS_KG
+                ),
+                LifterConstants.LENGTH_OF_LIFTER,
+                LifterConstants.BACKWARD_LIMIT.getRadians(),
+                LifterConstants.FORWARD_LIMIT.getRadians(),
+                false, LifterConstants.STARTING_ANGLE.getRadians()
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/simulationLifter/SimulationLifter.java:35]:
```suggestion
                false, 
                LifterConstants.STARTING_ANGLE.getRadians()
```

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java (hunk 20 above)
Comment (yonnbro) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java:23]: Do we even have limit switches on this mechanism?

**Review by yonnbro — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java (hunk 20 above)
Comment (greenblitz4590) [reply src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java:23]: Danna - yes, I asked

File: src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java [hunk 24]
```diff
@@ -9,65 +11,97 @@
...
+
+        motor.setSoftLimit(
+                CANSparkMax.SoftLimitDirection.kReverse,
+                LifterConstants.BACKWARD_LIMIT.getRadians()
+        );
+        motor.setSoftLimit(
+                CANSparkMax.SoftLimitDirection.kForward,
+                LifterConstants.FORWARD_LIMIT.getRadians()
+        );
+        motor.setIdleMode(CANSparkMax.IdleMode.kBrake);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/lifter/neoLifter/NeoLifter.java:36]: you created the idle contant but you dont use it..?

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

**Review by yonnbro — APPROVED** [review]
After Romi’s requests

**Review by bossyshrimp68 — APPROVED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

**Pull request merged**